File: python/Q3_longest_substring_without_repeating_characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#


# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left: int = 0
        result: int = 0
        char_set: set[str] = set()

        for right in range(len(s)):
            while s[right] in char_set:
                char_set.remove(s[left])
                left += 1

            char_set.add(s[right])

            if (new_length := right - left + 1) > result:
                result = new_length

        return result


# @lc code=end

File: rust/src/leetcode/Q3_longest_substring_without_repeating_characters.rs
/*
 * @lc app=leetcode id=3 lang=rust
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
use std::collections::HashSet;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let chars: Vec<char> = s.chars().collect();
        let mut current_set: HashSet<char> = HashSet::new();
        let mut left: usize = 0;
        let mut result: usize = 0;

        for right in 0..chars.len() {
            while current_set.contains(&chars[right]) {
                current_set.remove(&chars[left]);
                left += 1;
            }

            current_set.insert(chars[right]);

            if right - left + 1 > result {
                result = right - left + 1;
            }
        }

        result as i32
    }
}
// @lc code=end
